tiledb/sm/query/dimension_label/index_data.h

```diff
@@ -0,0 +1,160 @@
+/**
+ * @file index_data.h
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2022 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * Class for internally managed index data for dimension labels.
+ */
+
+#ifndef TILEDB_INDEX_DATA_H
+#define TILEDB_INDEX_DATA_H
+
+#include "tiledb/common/common.h"
+#include "tiledb/type/range/range.h"
+
+#include <vector>
+
+using namespace tiledb::common;
+
+namespace tiledb::sm {
+
+enum class Datatype : uint8_t;
+
+/** Base class for internally managed index data. */
+class IndexData {
+ public:
+  virtual ~IndexData() = default;
+
+  /** Returns the count, or number of stored variables.*/
+  virtual uint64_t count() const = 0;
+
+  /** Returns pointer to data. */
+  virtual void* data() = 0;
+
+  /** Returns pointer to total data size. */
+  virtual uint64_t* data_size() = 0;
```

Comment:
Why do we need this?

Reply:
Because it is used to set the size in the `QueryBuffer`. It needs to be lvalue that doesn't go out of scope while the Query exists.

Reply:
Thanks!